datastore/providers/pgvector_datastore.py

```diff
@@ -0,0 +1,122 @@
+
+    id = Column(String, primary_key=True)
+    document_id = Column(String)
+    text = Column(String)
+    embedding = Column(Vector(1536))
+
+    # Add a Cosine Distance index for faster querying
+    index = Index(
+        "vector_cosine_idx",
+        embedding,
+        postgresql_using="ivfflat",
+        postgresql_ops={"embedding": "vector_cosine_ops", "lists": "100"},
+    )
+
+
+class PgVectorDataStore(DataStore):
+    def __init__(self):
+        # Read the database URL from the PGVECTOR_URL environment variable
+        if not PGVECTOR_URL:
+            raise ValueError("PGVECTOR_URL environment variable is not set")
+
+        self.engine = create_engine(PGVECTOR_URL)
+        Base.metadata.create_all(bind=self.engine)
+        self.session_factory = sessionmaker(bind=self.engine)
+
+    async def _upsert(self, chunks: Dict[str, List[DocumentChunk]]) -> List[str]:
+        with self.session_factory() as session:
+            for document_id, document_chunks in chunks.items():
+                for chunk in document_chunks:
+                    vector_doc = VectorDocument(
+                        id=chunk.id,
+                        document_id=document_id,
+                        text=chunk.text,
+                        embedding=chunk.embedding,
+                    )
+                    session.merge(vector_doc)
+
+            session.commit()
+
+        return list(chunks.keys())
+
+    async def _query(self, queries: List[QueryWithEmbedding]) -> List[QueryResult]:
+        with self.session_factory() as session:
+            results = []
+            for query in queries:
+                query_embedding = query.embedding
+                stmt = (
+                    select(VectorDocument)
+                    .order_by(VectorDocument.embedding.cosine_distance(query_embedding))
+                    .limit(query.top_k)
+                )
+                matched_documents = session.execute(stmt)
+                matched_documents = matched_documents.scalars().all()
+
+                # Calculate cosine similarity from cosine distance
+                query_results = [
+                    {
+                        "document_id": doc.document_id,
+                        "text": doc.text,
+                        "score": 1 - doc.embedding.cosine_distance(query_embedding),
+                    }
+                    for doc in matched_documents
+                ]
+
+                results.append(QueryResult(query=query.query, results=query_results))
+
+            return results
+
+    async def delete(
+        self,
+        ids: Optional[List[str]] = None,
+        filter: Optional[DocumentMetadataFilter] = None,
+        delete_all: Optional[bool] = None,
+    ) -> bool:
+        with self.session_factory() as session:
+            stmt = select(VectorDocument)
+
+            if ids:
+                stmt = stmt.where(VectorDocument.id.in_(ids))
+
+            if filter and filter.document_id:
```

Comment:
this could be any kind of metadata filter, see the `DoucmentMetadata` type. You should support delete by any kind of filter.